File: beam-android/core/media/src/main/kotlin/dev/beam/android/core/media/player/ProgressReporter.kt
package dev.beam.android.core.media.player

import dev.beam.android.core.ffi.repository.PlaybackRepository
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

/**
 * Keeps the server's idea of where the viewer is roughly current.
 *
 * The throttle itself lives in the Rust core, shared with every other client,
 * so this deliberately does *not* decide when to send. It samples on a timer
 * and forces on the events where losing the position would be noticed, and the
 * core decides what actually goes over the wire.
 *
 * Forcing matters more than the interval. A viewer who pauses and closes the
 * app expects to resume where they stopped; if the last sample went out on the
 * interval, they lose up to a full interval of progress, which reads as the
 * app having forgotten. Pause, seek and stop are therefore always forced.
 */
internal class ProgressReporter(
    private val repository: PlaybackRepository,
    private val scope: CoroutineScope,
    private val sampleIntervalMs: Long = DEFAULT_SAMPLE_INTERVAL_MS,
) {
    private var ticker: Job? = null

    /** Begin sampling [position] until [stop] is called. */
    fun start(
        fileId: String,
        position: () -> PlaybackPosition?,
    ) {
        stop()
        ticker =
            scope.launch {
                while (isActive) {
                    delay(sampleIntervalMs)
                    val sample = position() ?: continue
                    report(fileId, sample, force = false)
                }
            }
    }

    /** Send a position now, bypassing the interval. */
    fun forceReport(
        fileId: String,
        sample: PlaybackPosition,
    ) {
        scope.launch { report(fileId, sample, force = true) }
    }

    /** Stop sampling. Does not send a final position; the caller forces that. */
    fun stop() {
        ticker?.cancel()
        ticker = null
    }

    private suspend fun report(
        fileId: String,
        sample: PlaybackPosition,
        force: Boolean,
    ) {
        // Failures are swallowed on purpose. The core queues an undeliverable
        // position durably and drains it later, so surfacing an error here
        // would interrupt playback for something already handled -- and there
        // is nothing the viewer could usefully do about it mid-title.
        runCatching {
            repository.reportProgress(
                fileId = fileId,
                positionSecs = sample.positionMs / MILLIS_PER_SECOND,
                durationSecs =
                    sample.durationMs
                        ?.takeIf { it > 0L }
                        ?.let { it / MILLIS_PER_SECOND },
                force = force,
            )
        }
    }

    internal companion object {
        /**
         * Matches `beam-web`'s `usePlaybackBeacon`, so a viewer moving between
         * the web app and the phone sees the same resume granularity rather
         * than two different answers to "where was I".
         */
        const val DEFAULT_SAMPLE_INTERVAL_MS: Long = 15_000L
        const val MILLIS_PER_SECOND: Double = 1_000.0
    }
}

/** A position sample, in milliseconds. */
internal data class PlaybackPosition(
    val positionMs: Long,
    val durationMs: Long?,
)

File: beam-apple/Packages/BeamKit/Sources/BeamPlayback/ProgressReporter.swift
import BeamCore
import BeamFFI
import Foundation

/// Samples playback position and hands it to the core.
///
/// Deliberately thin. The throttle, the coalescing and the durable retry queue
/// all live in `beam-client-core`'s `progress` module, so an interrupted
/// report survives a crash and behaves identically on Android. This only
/// decides *when to sample*, and forces a report at the moments where losing
/// one would be visible: a pause, a seek, the end of an item, and going away.
///
/// The 15-second interval matches `beam-web`'s `usePlaybackBeacon` and
/// `beam-android`'s `ProgressReporter`, so "continue watching" behaves the
/// same wherever the person was last watching.
@MainActor
public final class ProgressReporter {
    /// How often to sample while playing.
    public static let interval: Duration = .seconds(15)

    private let playback: any PlaybackRepository
    private var ticker: Task<Void, Never>?
    private var fileId: String?

    /// Report through `playback`.
    public init(playback: any PlaybackRepository) {
        self.playback = playback
    }

    deinit {
        ticker?.cancel()
    }

    /// Begin sampling for `fileId`.
    ///
    /// - Parameter position: a closure read on each tick, rather than a value
    ///   pushed in, so the reporter always sees the engine's current position
    ///   and can never send a stale one.
    public func start(fileId: String, position: @escaping @MainActor () -> (Double, Double?)) {
        stop()
        self.fileId = fileId
        ticker = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(for: Self.interval)
                guard !Task.isCancelled, let self else { return }
                let (seconds, duration) = position()
                await self.report(positionSeconds: seconds, durationSeconds: duration, force: false)
            }
        }
    }

    /// Send a report now, bypassing the throttle.
    ///
    /// Used for pause, seek, end of item and backgrounding -- the moments a
    /// dropped report is the difference between resuming where someone was and
    /// resuming up to fifteen seconds earlier.
    public func flush(positionSeconds: Double, durationSeconds: Double?) async {
        await report(
            positionSeconds: positionSeconds, durationSeconds: durationSeconds, force: true)
    }

    /// Stop sampling.
    public func stop() {
        ticker?.cancel()
        ticker = nil
    }

    private func report(positionSeconds: Double, durationSeconds: Double?, force: Bool) async {
        guard let fileId, positionSeconds.isFinite, positionSeconds >= 0 else { return }
        // A failed report is not shown to the viewer and not retried here: the
        // core has already queued it durably, and an error banner over a film
        // because a progress beacon missed would be worse than the miss.
        _ = try? await playback.reportProgress(
            fileId: fileId,
            positionSeconds: positionSeconds,
            durationSeconds: durationSeconds,
            force: force
        )
    }
}
